app/models/payment_gateway_setting.rb

```diff
@@ -37,4 +39,18 @@ def active_gateway_type
       errors.add(:gateway_type, :invalid)
     end
   end
+
+  def increment_failure_count
+    incremented_failure_count = failure_count + 1
+    self.gateway_settings = symbolized_settings.merge({failure_count:  incremented_failure_count})
```

Comment:
This doesn't look thread-safe.

Reply:
Maybe not an issue, but at least it can explain in case we once realize `FAILURE_THRESHOLD` is not being ensured.

Reply:
we are moving it to redis